### DISCUSSION TITLE
Summer not reading properties

### DISCUSSION THREAD

**Discussion by ThomasKennings** [JSBSim-Team/jsbsim#655] [Q&A]
Hello,

I have a summer component in my flight controls with a bias applied (code below). When I output fcs/target-alt-diff to a CSV, the output is just the bias value. Setting bias = 0 or removing the bias entirely makes all values of fcs/target-alt-diff 0. Using output from a simulink model, I know that the airplane's altitude is constantly changing, so it should not read 0. Replacing the property in question (ex: using velocities/vc-knots instead) does not fix the issue.

Does anyone have suggestions? Am I doing something wrong in presenting the property to the summer?

```xml	
<summer name="fcs/target-alt-diff">
    <property> position/h-agl-ft </property>
    <bias> -10000.0 </bias>
</summer>
```

**Comment by bcoconni** [comment]
It seems you forgot to include a code sample in your question ?

**Reply by ThomasKennings** [reply]
Woops. Updated with code. For some reason it isn't formatting like it should but it might be good enough.

**Reply by seanmcleod** [reply]
> For some reason it isn't formatting like it should

You must use triple back-ticks and then also ideally specify a language for code formatting, so I've edited it to be ```xml....

**Comment by seanmcleod** [comment]
Take a look at this example:

https://github.com/JSBSim-Team/jsbsim/blob/05c0b1762236395a539f6b1b9d82b250a4cc8a0d/aircraft/F450/FlightControl.xml#L13-L21

And this one:

https://github.com/JSBSim-Team/jsbsim/blob/a35be5ce6f7d7ca0ef9a11a62563e54d769bc4f4/aircraft/global5000/global5000apx.xml#L110-L113

**Reply by bcoconni** [reply]
To complement the examples shown by @seanmcleod, `<summer>` sums multiple `<input>` elements. So you need to supply at least 2 `<input>` elements. The `<bias>` factor is applied **after** the result of the summation is processed. So in your case, as there is only 1 `<input>` element, no summation is done (i.e. the single `<input>` element you provide is ignored) and `<bias>`is applied to the kind-of absence of summation hence the weird behavior.

**Reply by ThomasKennings** [reply]
1. I feel incredibly stupid. I was looking straight at all the <input> tags and my mind just went with the <property> tags I'd been seeing all day before!
2. It sounds like the summer is inappropriate for adding a value to a property's value, and storing that quantity in a new property. Is there some element I should use instead to try and do this? I'm trying to do PID control targeting a given altitude and intended to use summer to calculate the difference between the current and target altitude.

**Reply by seanmcleod** [reply]
> It sounds like the summer is inappropriate.... 

Hmm, seems as if you're still squinting too hard :wink:

The second example I listed above from the global 5000 auto-pilot is **exactly** what you're looking for, it has a heading set point, `ap/heading_setpoint` which it subtracts (by adding to a negated value, as mentioned in the documentation I posted below for the `summer` element) from the current heading `-fcs/heading-true-degrees` and outputs a property which contains the error, i.e. `fcs/heading-error`.

**Reply by ThomasKennings** [reply]
Was not on my game yesterday apparently! I will try that out.

**Comment by seanmcleod** [comment]
And the documentation:

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/src/models/flight_control/FGSummer.h#L54-L72

**Comment by seanmcleod** [comment]
@ThomasKennings I presume this work is for the following work at the University of Michigan? 

> Real-Time Computing Laboratory (RTCL) research into Boeing 737-MAX flight control algorithms.

Out of interest which 737 flight model are you using? The 737 model that is included in the JSBSim repo, or some other 737 model? I'm assuming if you started with the 737 model in the JSBSim repo that you've updated it to model things like the change in pitching moment from engine cowls at fairly large angles of attack, modelled the horizontal stabilizer etc.?

**Reply by ThomasKennings** [reply]
We are still using the default 737 model. We have not considered yet that those things could be modeled. If we can model those things, or if there are existing models of the 737-MAX out there, we ought to make sure we're getting that right.

Tagging my teammate/supervisor @noah-curran

**Comment by seanmcleod** [comment]
Ah, I "know" Noah in terms of answering some of his JSBSim Matlab S-Function questions recently :wink:
